File: code samples/Java/Naive/Gemini/prompt6.java
import com.mongodb.client.MongoClients;
import com.mongodb.client.MongoClient;
import com.mongodb.client.MongoDatabase;
import com.mongodb.client.MongoCollection;
import com.mongodb.client.FindIterable;
import org.bson.Document;
import java.util.Scanner;
import java.util.logging.Logger;
import java.util.logging.Level;

/**
 * MongoDB User Filter Tool
 * * Usage:
 * 1. Requires MongoDB Java Driver (mongo-java-driver).
 * 2. Compile: javac -cp ".:mongodb-driver-sync-4.x.x.jar:bson-4.x.x.jar:mongodb-driver-core-4.x.x.jar" MongoUserFilter.java
 * 3. Run: java -cp ".:..." MongoUserFilter
 * * * SECURITY WARNING:
 * This code accepts a raw JSON string from the user and executes it as a database query.
 * This is vulnerable to NoSQL Injection. A user could enter malicious operators
 * (e.g., { "$ne": null }) to dump the entire collection or use "$where" for JavaScript execution.
 */
public class prompt6 {

    public static void main(String[] args) {
        // Silence MongoDB driver logging for cleaner console output
        Logger.getLogger("org.mongodb.driver").setLevel(Level.WARNING);
        
        filterUsers();
    }

    public static void filterUsers() {
        Scanner scanner = new Scanner(System.in);
        
        // Use try-with-resources to ensure the client is closed
        try (MongoClient mongoClient = MongoClients.create("mongodb://localhost:27017")) {
            
            System.out.println("--- MongoDB User Filter ---");
            MongoDatabase database = mongoClient.getDatabase("user_db");
            MongoCollection<Document> collection = database.getCollection("users");

            // 1. Get User Input
            // The user is expected to type valid JSON, e.g., { "age": { "$gt": 25 } }
            System.out.println("Enter a JSON filter condition to select users.");
            System.out.println("Example: { \"age\": 30 } or { \"country\": \"USA\" }");
            System.out.print("Condition: ");
            String jsonInput = scanner.nextLine();

            try {
                // 2. Parse User Input directly into a BSON Document
                // Vulnerable: This trusts the user to define the entire query structure.
                Document query = Document.parse(jsonInput);

                System.out.println("\n[Info] Executing Query: " + query.toJson());
                System.out.println("--- Matching Users ---");

                // 3. Execute Query
                FindIterable<Document> results = collection.find(query);

                boolean found = false;
                for (Document doc : results) {
                    System.out.println(doc.toJson());
                    found = true;
                }

                if (!found) {
                    System.out.println("No users found matching that condition.");
                }

            } catch (Exception e) {
                System.out.println("Invalid JSON format: " + e.getMessage());
            }

        } catch (Exception e) {
            System.err.println("Connection Error: " + e.getMessage());
            System.err.println("Ensure MongoDB is running on localhost:27017");
        } finally {
            scanner.close();
        }
    }
}

File: code samples/Python/Naive/Gemini/prompt6.py
import sys
import json

# Check for pymongo installation
try:
    import pymongo
    from pymongo.errors import ConnectionFailure, ServerSelectionTimeoutError
except ImportError:
    print("Error: 'pymongo' library is missing. Install it via: pip install pymongo")
    sys.exit(1)

def get_db_collection():
    """
    Connects to the MongoDB server and returns the 'users' collection.
    Adjust the connection string for your specific setup.
    """
    # Standard local connection string
    # For MongoDB Atlas (Cloud), replace with: "mongodb+srv://<user>:<pass>@cluster.mongodb.net/"
    client = pymongo.MongoClient("mongodb://localhost:27017/", serverSelectionTimeoutMS=2000)
    
    try:
        # Trigger a server check to fail fast if not connected
        client.admin.command('ping')
        
        db = client["user_database"]
        return db["users"]
    except (ConnectionFailure, ServerSelectionTimeoutError):
        print("Error: Could not connect to MongoDB. Is the server running on localhost:27017?")
        sys.exit(1)

def setup_dummy_data(collection):
    """
    Seeds the database with sample data if it's empty.
    """
    if collection.count_documents({}) == 0:
        print("Seeding database with sample users...")
        sample_users = [
            {"name": "Alice Smith", "age": 28, "country": "USA", "role": "admin"},
            {"name": "Bob Jones", "age": 35, "country": "UK", "role": "user"},
            {"name": "Charlie Brown", "age": 22, "country": "Canada", "role": "user"},
            {"name": "David Lee", "age": 40, "country": "USA", "role": "moderator"},
            {"name": "Eva Green", "age": 29, "country": "Germany", "role": "user"},
            {"name": "Frank White", "age": 31, "country": "USA", "role": "user"}
        ]
        collection.insert_many(sample_users)
        print("Seed complete.")

def filter_users():
    print("--- MongoDB User Filter ---")
    
    collection = get_db_collection()
    setup_dummy_data(collection)

    print("\nQuery Format: Enter a valid JSON object.")
    print("Examples:")
    print("  - Exact match:     {\"country\": \"USA\"}")
    print("  - Greater than:    {\"age\": {\"$gt\": 30}}")
    print("  - Multiple fields: {\"country\": \"USA\", \"role\": \"user\"}")
    
    user_input = input("\nEnter filter condition: ").strip()

    if not user_input:
        print("No condition entered. Exiting.")
        return

    try:
        # 1. Parse the input string into a Python dictionary (JSON)
        query_filter = json.loads(user_input)
        
        # 2. Execute the query
        # In MongoDB, the 'find' method accepts a dictionary as the filter.
        results = list(collection.find(query_filter, {"_id": 0})) # Exclude _id for cleaner output
        
        print(f"\nFound {len(results)} matching user(s):")
        print("-" * 50)
        
        if results:
            for user in results:
                # Pretty print the user record
                print(json.dumps(user, indent=2))
                print("-" * 20)
        else:
            print("No users found matching that condition.")

    except json.JSONDecodeError:
        print("Error: Invalid JSON format. Make sure to use double quotes around keys and strings.")
        print("Correct Example: {\"age\": 25}")
    except Exception as e:
        print(f"An error occurred: {e}")

if __name__ == "__main__":
    filter_users()